File: dart/lib/src/domain/get_domain_owner.dart
import '../domain/get_domain_address.dart';
import '../nft/get_nft_owner.dart';
import '../rpc/rpc_client.dart';
import '../states/registry.dart';

/// Parameters for getting domain owner
class GetDomainOwnerParams {
  const GetDomainOwnerParams({
    required this.rpc,
    required this.domain,
  });

  /// The RPC client for blockchain interaction
  final RpcClient rpc;

  /// The domain name to get the owner for
  final String domain;
}

/// Retrieves the owner of the specified domain.
///
/// If the domain is tokenized, the NFT's owner is returned;
/// otherwise, the registry owner is returned.
///
/// This matches js-kit/src/domain/getDomainOwner.ts
///
/// [params] - Parameters containing RPC client and domain name
///
/// Returns the owner address of the domain
Future<String> getDomainOwner(GetDomainOwnerParams params) async {
  final domainResult = await getDomainAddress(GetDomainAddressParams(
    domain: params.domain,
  ));

  final results = await Future.wait([
    RegistryState.retrieve(params.rpc, domainResult.domainAddress),
    getNftOwner(GetNftOwnerParams(
      rpc: params.rpc,
      domainAddress: domainResult.domainAddress,
    )),
  ]);

  final registry = results[0] as RegistryState;
  final nftOwner = results[1] as String?;

  return nftOwner ?? registry.owner;
}

File: js-kit/src/domain/getDomainOwner.ts
import {
  GetAccountInfoApi,
  GetTokenLargestAccountsApi,
  Rpc,
} from "@solana/kit";

import { _parseSnsDomain } from "../utils/parseSnsDomain";
import { _getSnsDomainOwner } from "./getSnsDomainOwner";

/**
 * Parameters for retrieving a domain owner.
 *
 * @example
 * ```ts
 * const params: GetDomainOwnerParams = { rpc, domain: "example.sns" };
 * ```
 */
export interface GetDomainOwnerParams {
  /** RPC client. */
  rpc: Rpc<GetAccountInfoApi & GetTokenLargestAccountsApi>;
  /** Full `.sns` domain name. */
  domain: string;
}

/**
 * Retrieves the owner of the specified domain. If the domain is tokenized,
 * the NFT's owner is returned; otherwise, the registry owner is returned.
 *
 * @param params Domain owner retrieval parameters
 * @param params.rpc RPC client implementing account and token-largest-account APIs
 * @param params.domain Full `.sns` domain name
 * @returns The domain owner address.
 *
 * @example
 * ```ts
 * const owner = await getDomainOwner({ rpc, domain: "example.sns" });
 * ```
 */
export const getDomainOwner = async ({ rpc, domain }: GetDomainOwnerParams) => {
  const trimmedDomain = _parseSnsDomain(domain);
  return _getSnsDomainOwner({ rpc, domain: trimmedDomain });
};
